### DISCUSSION TITLE
How do primitives fit into the metamodel

### DISCUSSION THREAD

**Discussion by msche** [modelint/shlaer-mellor-metamodel#56] [Q&A]
I was wondering how primitives like integer, char, boolean, etc would fit into the metamodel

**Comment by modelint** [comment]
Hi Mark,

Thanks for this excellent question. The quick answer is that they don't. Type (user data type) definition is relegated to a domain distinct from the metamodel.

In [[TTM]](https://github.com/modelint/shlaer-mellor-metamodel/wiki/Resources#ttm-databases-types-and-the-relational-model-the-third-manifesto-cj-date-hugh-darwen----links-to-the-actual-book-as-a-downloadable-pdf) and other writings C.J. Date et al regard type theory to be orthogonal to relational theory. They specifically refer to nonscalar (relational) and scalar (user defined) types.

Nonscalar types are relational such as tuple or relation. From the metamodel perspective, such nonscalar types are very white-box. The metamodel action semantics can perform lots of relational magic on such types. Here we perform a union on two relation values, for example:

```
resolved contracts = completed + rejected
```

Scalar types, on the other hand, are int, string, or something arbitrarily complex like video image. The user will build up complex types from primitives, most likely using type inheritance and then define whatever operations are supported by a given type. A system of defining scalar types is handled in a distinct domain external to the metamodel.

Now the metamodel action semantics can flow both scalar and nonscalar values as inputs and outputs to actions. But scalar values must be manipulated with user defined or primitive operations as follows:

```
coordinates = xyvalue.Polar()
nice image = raw image.Process( algorithm: .amesh )
counter.Reset()
```

An advantage of this approach is that the metamodel isn’t concerned with how exactly you define your typing system. As long as a type is defined as a set of values and comes with supplied operations, the action semantics can handle it.

The other advantage is that the metamodel is greatly simplified without limiting the kinds of scalar types you have available.

I do plan to provide a type definition domain in a separate repo at some point. I expect I will just steal from existing work since type definition is a relatively mature topic to which I don’t have anything special to contribute. Or possibly someone else can do it.

– Leon

**Reply by msche** [reply]
Thanks for your answer, Leon. I need to digest it, but will come back if there are any more unclarities/questions regarding this.

**Reply by msche** [reply]
One more slightly related question regarding this. As I understand right now a attribute has as one of their properties a type. That type could be a `Class` or a `Type`. Would this mean that `Class` extends `Type` or vice versa?

**Reply by modelint** [reply]
I think there might be a misunderstanding here. Also I see something I need to update on the Class and Attribute subsystem.

First the update:

The Attribute class in v4.7.2 shows a relationship (R24) to an import Type class. Since then I've given more thought to the bridge to the Type Definition domain (which doesn't exist yet). So I will delete R24 and the imported class and change the refrential attribute to:

    Type : Type Name

This will just be a string, for convenience, representing the name of the Type to be defined in the other domain.

The Atrribute.Class attribute is just a reference to the Class that the Attribute belongs to via R20.

So, coming to your question, no, a Type cannot be a Class. An Attribute has a Type, period. The Type is a scalar (non-relational) type such as a primitive (Integer) or more likely something like Pressure, GPS coordinate, etc. Type type definition can be as complex as desired with as many user supported operations as desired, but all of them defined in the Type Definition domain.

Hope that helps.

**Reply by msche** [reply]
thanks for the clarification
